flight/PiOS/STM32F4xx/pios_fskdac.c

```diff
@@ -0,0 +1,352 @@
+	PIOS_DEBUG_Assert(cfg);
+
+	struct pios_fskdac_dev * fskdac_dev;
+
+	fskdac_dev = (struct pios_fskdac_dev *) PIOS_FSKDAC_alloc();
+	if (!fskdac_dev) return -1;
+
+	// Handle for the IRQ
+	g_fskdac_dev = fskdac_dev;
+
+
+	PIOS_DAC_COMMON_Init(PIOS_FSKDAC_DMA_irq_cb);
+
+	TIM_SetAutoreload(TIM6, PIOS_PERIPHERAL_APB1_CLOCK / DAC_SBS);
+
+	DMA_DeInit(DMA1_Stream5);
+	DMA_InitTypeDef DMA_InitStructure;	
+	DMA_InitStructure.DMA_Channel = DMA_Channel_7;  
+	DMA_InitStructure.DMA_PeripheralBaseAddr = (uint32_t)&DAC->DHR12R1;
+	DMA_InitStructure.DMA_Memory0BaseAddr = (uint32_t)&MARK_SAMPLES[0];
+	DMA_InitStructure.DMA_BufferSize = SAMPLES_PER_BIT;
+	DMA_InitStructure.DMA_PeripheralDataSize = DMA_PeripheralDataSize_HalfWord;
+	DMA_InitStructure.DMA_MemoryDataSize = DMA_MemoryDataSize_HalfWord;
+	DMA_InitStructure.DMA_DIR = DMA_DIR_MemoryToPeripheral;
+	DMA_InitStructure.DMA_PeripheralInc = DMA_PeripheralInc_Disable;
+	DMA_InitStructure.DMA_MemoryInc = DMA_MemoryInc_Enable;
+	DMA_InitStructure.DMA_Mode = DMA_Mode_Circular;
+	DMA_InitStructure.DMA_Priority = DMA_Priority_High;
+	DMA_InitStructure.DMA_FIFOMode = DMA_FIFOMode_Disable;
+	DMA_InitStructure.DMA_FIFOThreshold = DMA_FIFOThreshold_HalfFull;
+	DMA_InitStructure.DMA_MemoryBurst = DMA_MemoryBurst_Single;
+	DMA_InitStructure.DMA_PeripheralBurst = DMA_PeripheralBurst_Single;
+	DMA_Init(DMA1_Stream5, &DMA_InitStructure);
+
+	/* Configure double buffering */
+	DMA_DoubleBufferModeConfig(DMA1_Stream5, (uint32_t)&SPACE_SAMPLES[0], DMA_Memory_0);
+	DMA_DoubleBufferModeCmd(DMA1_Stream5, ENABLE);
+
+	fskdac_dev->tx_state = IDLE;
+
+	*fskdac_id = (uintptr_t) fskdac_dev;
+
+	/* Enable DMA1_Stream5 */
+	DMA_Cmd(DMA1_Stream5, ENABLE);
+
+	/* Enable DAC Channel1 */
+	DAC_Cmd(DAC_Channel_1, ENABLE);
+
+	/* Enable DMA for DAC Channel1 */
+	DAC_DMACmd(DAC_Channel_1, ENABLE);
+
+	DMA_ITConfig(DMA1_Stream5, DMA_IT_TC, DISABLE);
+
+	return 0;
+}
+
+
+static void PIOS_FSKDAC_TxStart(uintptr_t fskdac_id, uint16_t tx_bytes_avail)
+{
+	struct pios_fskdac_dev * fskdac_dev = (struct pios_fskdac_dev *)fskdac_id;
+
+	bool valid = PIOS_FSKDAC_validate(fskdac_dev);
+	PIOS_Assert(valid);
+	
+	/* Enable transfer complete interrupt for this stream */
+	/* Once it is called then the next byte will be loaded */
+	DMA_ITConfig(DMA1_Stream5, DMA_IT_HT, ENABLE);
+}
+
+static void PIOS_FSKDAC_RegisterTxCallback(uintptr_t fskdac_id, pios_com_callback tx_out_cb, uintptr_t context)
+{
+	struct pios_fskdac_dev * fskdac_dev = (struct pios_fskdac_dev *)fskdac_id;
+
+	bool valid = PIOS_FSKDAC_validate(fskdac_dev);
+	PIOS_Assert(valid);
+
+	/* 
+	 * Order is important in these assignments since ISR uses _cb
+	 * field to determine if it's ok to dereference _cb and _context
+	 */
+	fskdac_dev->tx_out_context = context;
+	fskdac_dev->tx_out_cb = tx_out_cb;
+}
+
+/**
+ * Schedule the next bit to be written by setting the address of the
+ * double buffer
+ */
+static void pios_fskdac_set_symbol(struct pios_fskdac_dev * fskdac_dev, uint8_t sym)
```

Comment:
I just looked at this section a little more. If you have a moment, a future PR which expanded the theoretic justification for this section would be most welcome. Perhaps just links to the documents you used to develop this would be enough to maintain it.